kubectl-volsync/cmd/migration.go

```diff
@@ -21,10 +21,75 @@ import (
 	"github.com/spf13/viper"
 	"k8s.io/kubectl/pkg/util/i18n"
 	"k8s.io/kubectl/pkg/util/templates"
+
+	volsyncv1alpha1 "github.com/backube/volsync/api/v1alpha1"
 )
 
 // MigrationRelationship defines the "type" of migration Relationships
-const MigrationRelationship RelationshipType = "migration"
+const MigrationRelationshipType RelationshipType = "migration"
+
+// migrationRelationship holds the config state for migration-type
+// relationships
+type migrationRelationship struct {
+	Relationship
+	data migrationRelationshipData
+}
+
+// migrationRelationshipData is the state that will be saved to the
+// relationship config file
+type migrationRelationshipData struct {
+	Version     int
+	Destination *migrationRelationshipDestination
+}
+
+type migrationRelationshipDestination struct {
+	// Cluster context name
+	Cluster string
+	// Namespace on destination cluster
+	Namespace string
+	// Name of PVC being replicated
+	PVCName string
+	// Name of the migrationDestination object
+	MDName string
+	// Name of Secret holding SSH keys
+	SSHKeyName string
+	// Parameters for the migrationDestination
+	Destination volsyncv1alpha1.ReplicationDestinationRsyncSpec
+}
+
+func (mr *migrationRelationship) Save() error {
+	mr.Set("data", mr.data)
+	// resource.Quantity doesn't properly encode, so we need to do it manually
+	if mr.data.Destination != nil && mr.data.Destination.Destination.Capacity != nil {
```

Comment:
The `mr.Set()` on L61 will recursively put the mr.data into the Viper instance. The individual lower-level Set calls are only necessary for structs that don't serialize correctly (e.g., `resource.Capacity`).
As an example:
https://github.com/JohnStrunk/volsync/blob/2d0e8e6c374f5364c359def47b6be8dd6ee31bad/kubectl-volsync/cmd/replication.go#L72-L82
The set call encodes all of `.data`, but since Capacity doesn't encode correctly for the source and destination, those are set specifically.

Note that resource.Capacity types do decode correctly, so no special handling is needed in the load function.

Reply:
Yeah. I tried with the below code (pre-existing)
```
func (mr *migrationRelationship) Save() error {
	mr.Set("data", mr.data)
	// resource.Quantity doesn't properly encode, so we need to do it manually
	if mr.data.Destination != nil && mr.data.Destination.Destination.Capacity != nil {
		mr.Set("data.destination.spec.Capacity", mr.data.Destination.Destination.Capacity.String())
```
that resulted in below o/p
**o/p 1:**
```
# cat /root/.volsync/v3.yaml 
data:
  destination:
    spec:
      capacity: 2Gi
id: f77a765f-f25e-4052-9678-06f7a4e66cfa
type: migration
```

If I only keep `mr.Set("data", mr.data)` and remove `mr.Set("data.destination.spec.Capacity", mr.data.Destination.Destination.Capacity.String())`, relationship data will be saved but capacity gets written in f`ormat: BinarySI`

o/p2:
```
data:
  version: 1
  destination:
    cluster: ""
    namespace: dest
    pvcname: volsync
    mdname: dest-volsync-migration-dest
    sshkeyname: ""
    destination:
      replicationdestinationvolumeoptions:
        copymethod: Direct
        capacity:
          format: BinarySI
        storageclassname: csi-hostpath-sc
        accessmodes:
        - ReadWriteOnce
        volumesnapshotclassname: null
        destinationpvc: null
      sshkeys: volsync-rsync-dest-src-dest-volsync-migration-dest
      servicetype: ClusterIP
      address: 10.96.188.200
      port: null
      path: null
      sshuser: null
id: 443ab627-f55f-40dd-9fa0-c108195a0f90
type: migration
```
Then I tried with ugly-looking multiple mr.set ()

**Observation:**
> mr.Set("data.destination.spec.Capacity", mr.data.Destination.Destination.Capacity.String()) 
   The above statement will override `mr.Set("data", mr.data)` and just writes `Capacity` under destination just like above **o/p 1**

Reply:
sure enough. Turns out Viper doesn't actually flatten the config until it's written, so the initial set establishes a value for "data", but not the subkeys. So when we set("data.some.path"), that destroys the initial value for "data".
It looks possible to get the intended behavior by serializing then deserializing between the initial set and the 2nd one, but that's really ugly.  I'll see what else I can come up with.

Reply:
Solution for this is in #116